pkg/kubemark/controller.go

```diff
@@ -0,0 +1,345 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/golang/glog"
+)
+
+const (
+	namespaceKubemark = "kubemark"
+	hollowNodeName    = "hollow-node"
+	nodeGroupLabel    = "autoscaling.k8s.io/nodegroup"
+)
+
+// KubemarkController is a simplified version of cloud provider for kubemark. It allows
+// to add and delete nodes from a kubemark cluster and introduces nodegroups
+// by applying labels to the kubemark's hollow-nodes.
+type KubemarkController struct {
+	nodeTemplate    *apiv1.ReplicationController
+	externalCluster externalCluster
+	kubemarkCluster kubemarkCluster
+}
+
+// externalCluster is used to communicate with the external cluster that hosts
+// kubemark, in order to be able to list, create and delete hollow nodes
+// by manipulating the replication controllers.
+type externalCluster struct {
+	rcLister  listersv1.ReplicationControllerLister
+	rcSynced  cache.InformerSynced
+	podLister listersv1.PodLister
+	podSynced cache.InformerSynced
+	client    kubeclient.Interface
+}
+
+// kubemarkCluster is used to delete nodes from kubemark cluster once their
+// respective replication controllers have been deleted and the nodes have
+// become unready. This is to cover for the fact that there is no proper cloud
+// provider for kubemark that would care for deleting the nodes.
+type kubemarkCluster struct {
+	client            kubeclient.Interface
+	nodeLister        listersv1.NodeLister
+	nodeSynced        cache.InformerSynced
+	nodesToDelete     map[string]bool
+	nodesToDeleteLock sync.Mutex
+}
+
+// NewKubemarkController creates KubemarkController using the provided clients to talk to external
+// and kubemark clusters.
+func NewKubemarkController(externalClient kubeclient.Interface, externalInformerFactory informers.SharedInformerFactory,
+	kubemarkClient kubeclient.Interface, kubemarkNodeInformer informersv1.NodeInformer) (*KubemarkController, error) {
+	rcInformer := externalInformerFactory.InformerFor(&apiv1.ReplicationController{}, newReplicationControllerInformer)
+	podInformer := externalInformerFactory.InformerFor(&apiv1.Pod{}, newPodInformer)
+	controller := &KubemarkController{
+		externalCluster: externalCluster{
+			rcLister:  listersv1.NewReplicationControllerLister(rcInformer.GetIndexer()),
+			rcSynced:  rcInformer.HasSynced,
+			podLister: listersv1.NewPodLister(podInformer.GetIndexer()),
+			podSynced: podInformer.HasSynced,
+			client:    externalClient,
+		},
+		kubemarkCluster: kubemarkCluster{
+			nodeLister:        kubemarkNodeInformer.Lister(),
+			nodeSynced:        kubemarkNodeInformer.Informer().HasSynced,
+			client:            kubemarkClient,
+			nodesToDelete:     make(map[string]bool),
+			nodesToDeleteLock: sync.Mutex{},
+		},
+	}
+
+	kubemarkNodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: controller.kubemarkCluster.removeUnneededNodes,
+	})
+
+	rand.Seed(time.Now().UTC().UnixNano())
```

Comment:
This seeds global generator, which can be seeded somewhere else by something. You can use your private `rand.Rand` generator, store it in the object and seed it here.